### ISSUE TITLE
Fix updated msjunit/asm/uint32div.js after r25103

### ISSUE THREAD

**Issue by paul99** [v8mips/v8mips#106]
uint32div.js fails when divisor is zero. MIPS divu instruction returns undefined value on divide by 0. 

Both ARM cpus return 0 in this case (though that seems to be settable via SCTLR.DZ bit, it appears that they are relying on this case.

I don't see how this works on x64, as it appears to raise exception for divide by zero, but the code looks as if they just get 0 result.

I have an unpublished CL https://codereview.chromium.org/701063002, can you guys PTAL?

**Comment by paul99** [comment]
So first, please sanity check that there is not some simpler way to do this than a runtime check on every divide.

Several issues to consider:
- I did not update the 2-input Divu macro-asm function. It does not appear to be used, but that is probably not a good excuse.
- We may also have Mod functions that should be changed, I did not check. We don't have a failing test there.
- Maybe our unit tests should also be updated to test this?
- I used Movz which generates very bad code on r6. We should investigate `sell` instructions for this.
- The TF use case (below) overwrote the input divisor register, so I had to use a temp `at` in that case.
- We should change the simulator to return a deterministic/recognizable value for div-by-0 (lets do this after the refactoring for function size)
  
  ```
  0x46e144ec   204  00021043       sra     v0, v0, 1
  0x46e144f0   208  8fc3fff4       lw      v1, -12(fp)
  0x46e144f4   212  0062001b       divu    v1, v0
  0x46e144f8   216  00001012       mflo    v0
  0x46e144fc   220  00421821       addu    v1, v0, v0
  0x46e14500   224  0062a026       xor     s4, v1, v0
  0x46e14504   228  00629826       xor     s3, v1, v0
  0x46e14508   232  0274a024       and     s4, s3, s4
  0x46e1450c   236  0680003a       bltz    s4, 58
  0x46e14510   240  00000000       nop
  ```

**Comment by kilvadyb** [comment]
On first view I would exted Paul's fix for all the div, divu, mod, modu ops. but on ia32 the div, udiv also drop exception for 0 divisor:
http://x86.renejeschke.de/html/file_module_x86_id_72.html
and I have not found any check or fix in either assembler-ia32 nor macro-assembler-ia32 nor code-generator-ia32.cc. In lithium they handle 0 divisor specially for example:
  // Check for x % 0, idiv would signal a divide error. We have to
  // deopt in this case because we can't return a NaN.
x / 0 div,idiv also.

**Comment by kilvadyb** [comment]
On ia32 some pre-checking is compiled in (B14) so if the divisor is zero then returns Smi(zero):

```
                  -- B14 start --
0x55e583e4   100  83f900         cmp ecx,0x0
0x55e583e7   103  0f848a000000   jz 247  (0x55e58477)
                  -- B10 start --
0x55e583ed   109  f645f401       test_b [ebp-0xc],0x1
0x55e583f1   113  0f8559000000   jnz 208  (0x55e58450)
                  -- B13 start --
0x55e583f7   119  8b45f4         mov eax,[ebp-0xc]
0x55e583fa   122  d1f8           sar eax,1
                  -- B11 start --
0x55e583fc   124  cc             int3  ; I added this int to debug if this part is called at all - and not.
0x55e583fd   125  33d2           xor edx,edx
0x55e583ff   127  f7f1           div ecx
```

**Comment by dusmil** [comment]
Could we handle this in instruction selection phase?
I think it is better way than clobbering Divu macro, and make it heavy weight.

**Comment by kilvadyb** [comment]
Sorry for closing the issue mistakenly.

I agree. I am checking the turbofan code of ARM and if they have a chek-if-zero BXX then definitely we should move to that direction also instead of extending the DivU, ModU, ... macros.

**Comment by kilvadyb** [comment]
ARM does not have "if (divisor == 0)" check. In the turbofan graph of uint32div.js the highest Block is B13 while on ia32 the "if (divisor == 0)" is in B14. So extending the DivU, ... macros is a quick solution and following the ia32 way seems to be more polished. I can't see if there is any usage of Div, DivU, Mod, ModU macros in turbofan where we can avoid the "if (divisor == 0)" check. If not then easier to extend the macros or we can add an enum parameter to those macro codes:
enum { DIV_CHECK, DIV_NO_CHECK }; and the additional parameter is defaulted to DIV_CHECK.

**Comment by paul99** [comment]
So my take is ARM does not need this 0-check, as the arm udiv instruction returns 0 for a 0-divisor. But that's not true for ia32/x64, so they need the check. But the check is _not_ in InstructionSelector::VisitUint32Div(). So somehow they have knowledge that this 0-check is required at a higher level, and I have not seen how yet.

Also, Balázs, could you post your TF graph, I've been running debug ia32 & x64, and I am not seeing the same thing you mentioned above. Maybe you can post the file on slack, or put it in dropbox.

**Comment by dusmil** [comment]
Mistery solved.
Check InstructionSelector::SupportedMachineOperatorFlags() and related commit where we falsely say those ops. are safe.

**Comment by paul99** [comment]
Thanks Dusan. I deleted the old CL, created a new one, with kNoFlags, and it is landed as f3b633cc0 {#25174}
https://codereview.chromium.org/704013002

The mips32 sim builders at google & imgtec are green now. 

Thanks to both of you guys for pushing back on my bad CL, to make sure we found the right solution (I know it did not "feel right" to me, since I knew x86 did not handle div-by-zero)

A small puzzle still: in the Arm v8 manual is this: NOTE: The divide instructions do not generate a trap upon division by zero, but write zero to the destination register.

So why is arm64 not considered "safe"?

Still, this is solved now for MIPS. (You will need this in mips64 TF.) Closing.
